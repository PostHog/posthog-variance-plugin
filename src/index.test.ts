import type { PluginEvent } from '@posthog/plugin-scaffold'
import { getMeta, resetMeta } from '@posthog/plugin-scaffold/test/utils'
import fetchMock, { enableFetchMocks } from 'jest-fetch-mock'

import type { VarianceConfig } from '.'
import { VarianceEventType } from '.'
import autocapture from './fixtures/autocapture.json'
import alias from './fixtures/create_alias.json'
import custom from './fixtures/custom.json'
import feature_flag_called from './fixtures/feature_flag_called.json'
import groupidentify from './fixtures/groupidentify.json'
import identify from './fixtures/identify.json'
import pageleave from './fixtures/pageleave.json'
import pageview from './fixtures/pageview.json'
import { onEvent } from './index'

const config: VarianceConfig = {
  authHeader: `Basic ...`,
  webhookUrl: `https://variance.variance.dev`,
}
resetMeta({ config })

function buildEvent(json: Record<string, unknown>): PluginEvent {
  return {
    ...json,
    timestamp: `2020-11-26T12:58:58.453Z`,
    uuid: `0`,
  } as PluginEvent
}

describe(`supported`, () => {
  enableFetchMocks()
  afterEach(() => {
    fetchMock.resetMocks()
  })

  async function assertFetch(
    json: Record<string, unknown>,
    eventType: VarianceEventType
  ) {
    await onEvent(buildEvent(json), getMeta())
    expect(fetchMock.mock.calls[0][0]).toBe(config.webhookUrl)
    const opts = fetchMock.mock.calls[0][1]
    expect(opts?.headers).toEqual({
      'Authorization': config.authHeader,
      'Content-Type': `application/json`,
    })
    if (typeof opts?.body !== `string`) throw new Error(`Body should be string`)
    const body = JSON.parse(opts.body)
    expect(body.type).toBe(eventType)
    expect(body.messageId).toBe(`0`)
    expect(body).toMatchSnapshot()
  }

  it(`$create_alias`, async () => {
    await assertFetch(alias, VarianceEventType.alias)
  })

  it(`$groupidentify`, async () => {
    await assertFetch(groupidentify, VarianceEventType.group)
  })

  it(`$identify`, async () => {
    await assertFetch(identify, VarianceEventType.identify)
  })

  it(`$pageview`, async () => {
    await assertFetch(pageview, VarianceEventType.page)
  })

  it(`custom event`, async () => {
    await assertFetch(custom, VarianceEventType.track)
  })
})

describe(`ignore`, () => {
  const mockDebug = jest.spyOn(console, `debug`).mockImplementation(() => {
    /**/
  })

  afterEach(() => {
    mockDebug.mockReset()
  })

  async function assertIgnore(json: Record<string, unknown>) {
    await onEvent(buildEvent(json), getMeta())
    expect(mockDebug).toBeCalledWith(`Unsupported event: ${String(json.event)}`)
  }

  it(`$autocapture`, async () => {
    await assertIgnore(autocapture)
  })

  it(`$feature_flag_called`, async () => {
    await assertIgnore(feature_flag_called)
  })

  it(`$pageleave`, async () => {
    await assertIgnore(pageleave)
  })
})
