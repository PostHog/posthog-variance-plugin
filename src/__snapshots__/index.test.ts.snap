// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`supported $create_alias 1`] = `
{
  "anonymousId": "180d9a2c070861-0c82a7dd77400b-34736704-16a7f0-180d9a2c0711cbb",
  "context": {
    "app": {
      "version": "1.35.0",
    },
    "os": {
      "name": "Mac OS X",
    },
    "page": {
      "host": "localhost:8000",
      "path": "/events",
      "referrer": "http://localhost:8000/authorize_and_redirect/?redirect=http%3A%2F%2Flocalhost%3A8000%2Fproject%2Fsettings",
      "search": "",
      "url": "http://localhost:8000/events",
    },
    "screen": {
      "height": 982,
      "width": 1512,
    },
  },
  "libary": {
    "name": "posthog-variance-plugin",
    "version": "0.0.0",
  },
  "messageId": "0",
  "previousId": "YiZpN3z6sbgGuZ93bWpAqsfyaQDj1la4us2p3mT41uj",
  "timestamp": "2020-11-26T12:58:58.453Z",
  "type": "alias",
  "userId": "1",
}
`;

exports[`supported $groupidentify 1`] = `
{
  "anonymousId": "180d9a2c070861-0c82a7dd77400b-34736704-16a7f0-180d9a2c0711cbb",
  "context": {
    "app": {
      "version": "1.35.0",
    },
    "os": {
      "name": "Mac OS X",
    },
    "page": {
      "host": "localhost:8000",
      "path": "/project/apps",
      "referrer": "http://localhost:8000/authorize_and_redirect/?redirect=http%3A%2F%2Flocalhost%3A8000%2Fproject%2Fsettings",
      "search": "",
      "url": "http://localhost:8000/project/apps",
    },
    "screen": {
      "height": 982,
      "width": 1512,
    },
  },
  "groupId": "http://localhost:8000",
  "libary": {
    "name": "posthog-variance-plugin",
    "version": "0.0.0",
  },
  "messageId": "0",
  "timestamp": "2020-11-26T12:58:58.453Z",
  "traits": {
    "instance": "http://localhost:8000",
    "organization": "0180d9a2-b90a-0000-34d4-e261a7db0061",
    "posthog_group_type": "instance",
    "project": "0180d9a2-bdc7-0000-33c9-e4c7337dc44f",
  },
  "type": "group",
  "userId": "YiZpN3z6sbgGuZ93bWpAqsfyaQDj1la4us2p3mT41uj",
}
`;

exports[`supported $identify 1`] = `
{
  "libary": {
    "name": "posthog-variance-plugin",
    "version": "0.0.0",
  },
  "messageId": "0",
  "timestamp": "2020-11-26T12:58:58.453Z",
  "traits": {
    "email": "guillaume@variance.com",
  },
  "type": "identify",
  "userId": "0",
}
`;

exports[`supported $pageview 1`] = `
{
  "anonymousId": "180d9a2c070861-0c82a7dd77400b-34736704-16a7f0-180d9a2c0711cbb",
  "context": {
    "app": {
      "version": "1.35.0",
    },
    "os": {
      "name": "Mac OS X",
    },
    "page": {
      "host": "localhost:8000",
      "path": "/home",
      "referrer": "http://localhost:8000/authorize_and_redirect/?redirect=http%3A%2F%2Flocalhost%3A8000%2Fproject%2Fsettings",
      "search": "",
      "url": "http://localhost:8000/home",
    },
    "screen": {
      "height": 982,
      "width": 1512,
    },
  },
  "libary": {
    "name": "posthog-variance-plugin",
    "version": "0.0.0",
  },
  "messageId": "0",
  "properties": {
    "host": "localhost:8000",
    "path": "/home",
    "referrer": "http://localhost:8000/authorize_and_redirect/?redirect=http%3A%2F%2Flocalhost%3A8000%2Fproject%2Fsettings",
    "url": "http://localhost:8000/home",
  },
  "timestamp": "2020-11-26T12:58:58.453Z",
  "type": "page",
  "userId": "YiZpN3z6sbgGuZ93bWpAqsfyaQDj1la4us2p3mT41uj",
}
`;

exports[`supported custom event 1`] = `
{
  "anonymousId": "180d9a2c070861-0c82a7dd77400b-34736704-16a7f0-180d9a2c0711cbb",
  "context": {
    "os": {
      "name": "Mac OS X",
    },
    "page": {
      "host": "localhost:8000",
      "path": "/project/apps",
      "referrer": "http://localhost:8000/authorize_and_redirect/?redirect=http%3A%2F%2Flocalhost%3A8000%2Fproject%2Fsettings",
      "search": "",
      "url": "http://localhost:8000/project/apps",
    },
    "screen": {
      "height": 982,
      "width": 1512,
    },
  },
  "event": "team has ingested events",
  "libary": {
    "name": "posthog-variance-plugin",
    "version": "0.0.0",
  },
  "messageId": "0",
  "properties": {
    "distinct_id": "YiZpN3z6sbgGuZ93bWpAqsfyaQDj1la4us2p3mT41uj",
    "is_demo_project": false,
    "token": "phc_r06aTsEyFENJMKoqoJCoPxCtRCQGZFw1njECfKiup7K",
  },
  "timestamp": "2020-11-26T12:58:58.453Z",
  "type": "track",
  "userId": "YiZpN3z6sbgGuZ93bWpAqsfyaQDj1la4us2p3mT41uj",
}
`;
